src/store/index-level.ts

```diff
@@ -1,213 +1,284 @@
+  private async findExactMatches(
+    tenant:string,
+    propertyName: string,
+    propertyValue: unknown,
+    sortProperty: string,
+    sortDirection: SortOrder,
+    cursor?: string,
+    options?: IndexLevelOptions
+  ): Promise<IndexedItem<T>[]> {
+    const tenantPartition = await this.db.partition(tenant);
+    const indexPartition = await tenantPartition.partition(INDEX_SUBLEVEL_NAME);
 
+    const matchPrefix = this.constructIndexedKey(sortProperty, propertyName, this.encodeValue(propertyValue));
     const iteratorOptions: LevelWrapperIteratorOptions<string> = {
-      gt: propertyValuePrefix
+      gt: matchPrefix
     };
 
-    const matches: string[] = [];
-    for await (const [ key, dataId ] of partition.iterator(iteratorOptions, options)) {
-      if (!key.startsWith(propertyValuePrefix)) {
+    // if a cursor is defined we want to set it as the starting point for the query.
+    if (cursor !== undefined) {
+      iteratorOptions.gt = cursor;
+    }
+
+    const matches: IndexedItem<T>[] = [];
+    for await (const [ key, value ] of indexPartition.iterator(iteratorOptions, options)) {
+      // immediately stop if we arrive at an index that contains a different property
+      if (!key.startsWith(matchPrefix)) {
         break;
       }
 
-      matches.push(dataId);
+      const dataKey = this.extractIndexedKey(key);
+      matches.push({ key: dataKey, value: JSON.parse(value) });
+    }
+
+    if (sortDirection !== SortOrder.Ascending) {
+      return matches.reverse();
     }
+
     return matches;
   }
 
-  /**
-   * @returns IDs of data that matches the range filter.
-   */
-  private async findRangeMatches(tenant: string, propertyName: string, rangeFilter: RangeFilter, options?: IndexLevelOptions): Promise<string[]> {
-    const partition = await executeUnlessAborted(this.db.partition(tenant), options?.signal);
+  private async findRangeMatches(
+    tenant: string,
+    propertyName: string,
+    rangeFilter: RangeFilter,
+    sortProperty: string,
+    sortDirection: SortOrder,
+    cursor?: string,
+    options?: IndexLevelOptions
+  ): Promise<IndexedItem<T>[]> {
+    const tenantPartition = await this.db.partition(tenant);
+    const indexPartition = await tenantPartition.partition(INDEX_SUBLEVEL_NAME);
     const iteratorOptions: LevelWrapperIteratorOptions<string> = {};
+    const matchPrefix = this.constructIndexedKey(sortProperty, propertyName);
 
     for (const comparator in rangeFilter) {
       const comparatorName = comparator as keyof RangeFilter;
-      iteratorOptions[comparatorName] = this.join(propertyName, this.encodeValue(rangeFilter[comparatorName]));
+      iteratorOptions[comparatorName] = this.constructIndexedKey(sortProperty, propertyName, this.encodeValue(rangeFilter[comparatorName]));
+    }
+
+    // if a cursor exists, it will be the starting point for the range query but not equal to.
+    if (cursor !== undefined) {
```

Comment:
Move this check to top of function

Reply:
Moving it up top would mean adding more complexity/less understood code where the iteratorOptions are built above it.

Since it's modifying the starting point of the iterator if a cursor exists, it makes more sense to leave it here.